test-new/utils/server.js

```diff
@@ -0,0 +1,390 @@
+import http from 'http';
+import zlib from 'zlib';
+import parted from 'parted';
+const {multipart: Multipart} = parted;
+
+export default class TestServer {
+	constructor() {
+		this.server = http.createServer(this.router);
+		this.port = 30001;
+		this.hostname = 'localhost';
+		// Node 8 default keepalive timeout is 5000ms
+		// make it shorter here as we want to close server quickly at the end of tests
+		this.server.keepAliveTimeout = 1000;
+		this.server.on('error', err => {
+			console.log(err.stack);
+		});
+		this.server.on('connection', socket => {
+			socket.setTimeout(1500);
+		});
+	}
+
+	start(cb) {
+		this.server.listen(this.port, this.hostname, cb);
+	}
+
+	stop(cb) {
+		this.server.close(cb);
+	}
+
+	mockResponse(responseHandler) {
+		this.server.nextResponseHandler = responseHandler;
+		return `http://${this.hostname}:${this.port}/mocked`;
+	}
+
+	router(request, response) {
```

Comment:
Something to think about for the future, in Got, we ended up with a separate server for each test, which is really nice as you don't need a router and nothing can tests cannot affect each other. We also use a helper and AVA test macros to make it really convenient: https://github.com/sindresorhus/got/blob/a337dfdfc4151593044c8b729245308081a071be/test/http.ts#L9-L15

Reply:
That's really interesting, will definitely look into it. Thanks :smile: